### ISSUE TITLE
integration of full-text-search into readium cloud reader

### ISSUE THREAD

**Issue by hajarghaem** [larsvoigt/epub-full-text-search#19]
@larsvoigt
Dear Lars,
I have tried to integrate full-text-search in my Readium cloud reader, by flowing the instructions you shared in https://github.com/readium/readium-js/issues/17:
"epub-full-text-search usage
If you want see how this feature can be implemented then check out branch.
To get it run call:
Prerequisites are same like the orginal readium repro
npm run build
npm run dist:cloudReaderWithFullTextSearch"
But I cannot integrate it in my Cloud reader that locally works on my computer. When I use "npm run dist:cloudReaderWithFullTextSearch", my build-output folder is empty. Also the dist folder does not consist of all required folders and files.
Could you please let me know if I need to take into account specific points.
Thanks in advance,
Hajar

**Comment by larsvoigt** [comment]
Hello @hajarghaem, 

the full-text-search feature is not a part of the main readium cloud reader project. So the only way to test it is, you can check out this [branch](https://github.com/larsvoigt/readium-js-viewer/blob/feature/new_full_text_search).  I have added some short instructions to the [README](https://github.com/larsvoigt/readium-js-viewer/blob/feature/new_full_text_search/README.md) how you can build it. I hope it helps a little bit otherwise contact me agian :-). 
Lars

**Comment by hajarghaem** [comment]
Hi Lars,

Thanks for your time, updating the branch, and response :).

I was following you in the last one hour ;).

I did not do step 1: install epub-full-text-search
<https://github.com/larsvoigt/epub-full-text-search>.

What I want is :

1- I have a Readium Cloud Reader.

2- I want to add the search function (using your full-text-search feature)
as part of functionality of my Readium Cloud Reader (Something like
annotations, or other button on the top right of Readium)

3- As far as I understand from you previous email, I am not able to
integrate your full-text-search feature, as part of my Readium Cloud
Reader! right?

4-Could you please let me know your opinion/suggestion for the best way to
add this feature to Readium could reader?

5- Regarding the installation instruction you said: I did not do step 1,
but I did other steps. After step 3: the build-output is empty. After step
4: the dist folder has just index.html and script folder that does not
include any JavaScript files.

Thanks in advance,
Kind regards,
Hajar

**Comment by larsvoigt** [comment]
I see! 
Do you have checked out the whole readium cloudreader repository with all subs modules ...? 
 If so  you can try to merge my branch into yours. 
And then:  
```npm run build``` 
``` npm run dist:cloudReader ``` not more ```npm run dist:cloudReaderWithFullTextSearch``` .

After the dist folder should be full filled.

**Comment by larsvoigt** [comment]
Maybe ths [guidance](http://blog.xk72.com/post/42808949385/merge-in-another-git-repository) can help to merge a branch form repository to an other.

**Comment by hajarghaem** [comment]
Hi Lars,
Thanks!
But my problems are still there.
dist:cloudReader does not have required files, and I fill it with the files I had form my another Readium Cloud reader folder. Also the build-output is empty.
As you said, your current Full-Text-Search cannot be used as part of Readium Cloud Reader!
And I my goal is integrating it as a feature of Readium. Could you please help me with your opinion about possible ways to integrate search feature into Readium?

Thanks in advance,
Hajar

**Comment by larsvoigt** [comment]
Hi Hajar,
how you build your cloudreader?

Okay try this:

1. Have you done all steps form [orginal readium repro](https://github.com/readium/readium-js-viewer/tree/develop#development) ? 

2. Merge or copy all my [changes (open "Files changed"-tab)](https://github.com/readium/readium-js-viewer/compare/master...larsvoigt:feature/new_full_text_search) to your cloud reader repository and you have this feature included. ( ignore readme)

3. Then ```npm run build```  will bundle all readium js-files and libs to one bundle file.   Can you do this?  

4. Then try ``` npm run dist:cloudReader ```  that will bundle all necessary files for the cloudreader to the dist-folder. 

That it ...
Lars

**Comment by hajarghaem** [comment]
what I did is:
1- git clone git clone --recursive -b develop
https://github.com/*readium*/readium-js-viewer.git
readium-js-viewer I als tried one time with ( git clone git clone
--recursive -b feature/new-full-text-search
https://github.com/larsvoigt/readium-js-viewer.git)
2- cd readium-js-viewer
3- and I did other steps in
https://github.com/readium/readium-js-viewer/tree/develop#development but
not "npm run prepare"
4- then I copy pasted your changed files from
https://github.com/readium/readium-js-viewer/compare/master...larsvoigt:feature/new_full_text_search
5- npm run prepare
6- npm run build
7- npm run dist:cloudReader

But still the dist/cloud-reader folder has just two empty folder and not
other files. The build-output folder is empty

Please let me know what is wrong with my steps?
Thanks in advance,
Hajar

**Comment by larsvoigt** [comment]
Any error messages if you run the task scripts?

**Comment by hajarghaem** [comment]
I did one more time the following steps:
1- git clone --recursive -b develop https://github.com/*readium*/readium-js-viewer.git readium-js-viewer 2- cd readium-js-viewer 
3- and I did other steps in https://github.com/readium/readium-js-viewer/tree/develop#development but not "npm run prepare"
 4- then I copy pasted your changed files from readium/readium-js-viewer@master...larsvoigt:feature/new_full_text_search
 5- npm run prepare 
6- npm run build 
7- npm run dist:cloudReader.
This time the build-output folder is includes: "multiple-bundels" and "XXX" folders with sub-folders and files. But the dist/cloud-reader folder still  has two empty folder and not other files. 
I have error when running "npm run dist:cloudReader". I should say that my node use older version because of https://github.com/readium/readium-js-viewer/issues/509 and https://github.com/readium/readium-js-viewer/issues/453#issuecomment-169410695.
When I use "npm run dist" in the original Readium instruction (ot including your changed fles), then creates dist with its sub-folders and files.



My logs:
$ npm run dist:cloudReader

> readium-js-viewer@0.24.0 predist:cloudReader C:\Users\Hajar\Projects\ReadiumCloud\test7\readium-js-viewer
> node ./readium-js/readium-shared-js/readium-cfi-js/node_modules/rimraf/bin.js dist/cloud-reader && node ./readium-js/readium-shared-js/readium-cfi-js/node_modules/rimraf/bin.js dist/cloud-reader_sourcemap


> readium-js-viewer@0.24.0 dist:cloudReader C:\Users\Hajar\Projects\ReadiumCloud\test7\readium-js-viewer
> npm run dist:cloudReader:copy && npm run dist:cloudReader:css && npm run dist:cloudReader:js


> readium-js-viewer@0.24.0 dist:cloudReader:copy C:\Users\Hajar\Projects\ReadiumCloud\test7\readium-js-viewer
> node ./readium-js/readium-shared-js/readium-cfi-js/node_modules/mkdirp/bin/cmd.js dist/cloud-reader/scripts/zip && node ./readium-js/readium-shared-js/readium-cfi-js/node_modules/mkdirp/bin/cmd.js dist/cloud-reader/scripts/mathjax && node ./readium-js/readium-shared-js/readium-cfi-js/node_modules/cpy-cli/cli.js src/cloud-reader/index.html dist/cloud-reader/ && node ./readium-js/readium-shared-js/readium-cfi-js/node_modules/cpy-cli/cli.js build-output/z-worker.js dist/cloud-reader/scripts/zip && node ./readium-js/readium-shared-js/readium-cfi-js/node_modules/cpy-cli/cli.js build-output/deflate.js dist/cloud-reader/scripts/zip && node ./readium-js/readium-shared-js/readium-cfi-js/node_modules/cpy-cli/cli.js build-output/inflate.js dist/cloud-reader/scripts/zip && node ./readium-js/readium-shared-js/readium-cfi-js/node_modules/ncp/bin/ncp src/images/ dist/cloud-reader/images/ && node ./readium-js/readium-shared-js/readium-cfi-js/node_modules/cpy-cli/cli.js node_modules/mathjax-single-file/dist/MMLSVG/MathJax.js dist/cloud-reader/scripts/mathjax && node ./readium-js/readium-shared-js/readium-cfi-js/node_modules/ncp/bin/ncp node_modules/bootstrap/fonts/ dist/cloud-reader/fonts/ && node ./readium-js/readium-shared-js/readium-cfi-js/node_modules/cpy-cli/cli.js src/css/annotations.css dist/cloud-reader/css/

Cannot glob `build-output/z-worker.js`: Cannot copy from `C:\Users\Hajar\Projects\ReadiumCloud\test7\readium-js-viewer\build-output\z-worker.js` to `C:\Users\Hajar\Projects\ReadiumCloud\test7\readium-js-viewer\dist\cloud-reader\scripts\zip\z-worker.js`: cannot stat path `C:\Users\Hajar\Projects\ReadiumCloud\test7\readium-js-viewer\build-output\z-worker.js`: ENOENT: no such file or directory, stat 'C:\Users\Hajar\Projects\ReadiumCloud\test7\readium-js-viewer\build-output\z-worker.js'

npm ERR! Windows_NT 10.0.14393
npm ERR! argv "C:\\Program Files\\nodejs\\node.exe" "C:\\Program Files\\nodejs\\node_modules\\npm\\bin\\npm-cli.js" "run" "dist:cloudReader:copy"
npm ERR! node v6.1.0
npm ERR! npm  v3.8.6
npm ERR! code ELIFECYCLE
npm ERR! readium-js-viewer@0.24.0 dist:cloudReader:copy: `node ./readium-js/readium-shared-js/readium-cfi-js/node_modules/mkdirp/bin/cmd.js dist/cloud-reader/scripts/zip && node ./readium-js/readium-shared-js/readium-cfi-js/node_modules/mkdirp/bin/cmd.js dist/cloud-reader/scripts/mathjax && node ./readium-js/readium-shared-js/readium-cfi-js/node_modules/cpy-cli/cli.js src/cloud-reader/index.html dist/cloud-reader/ && node ./readium-js/readium-shared-js/readium-cfi-js/node_modules/cpy-cli/cli.js build-output/z-worker.js dist/cloud-reader/scripts/zip && node ./readium-js/readium-shared-js/readium-cfi-js/node_modules/cpy-cli/cli.js build-output/deflate.js dist/cloud-reader/scripts/zip && node ./readium-js/readium-shared-js/readium-cfi-js/node_modules/cpy-cli/cli.js build-output/inflate.js dist/cloud-reader/scripts/zip && node ./readium-js/readium-shared-js/readium-cfi-js/node_modules/ncp/bin/ncp src/images/ dist/cloud-reader/images/ && node ./readium-js/readium-shared-js/readium-cfi-js/node_modules/cpy-cli/cli.js node_modules/mathjax-single-file/dist/MMLSVG/MathJax.js dist/cloud-reader/scripts/mathjax && node ./readium-js/readium-shared-js/readium-cfi-js/node_modules/ncp/bin/ncp node_modules/bootstrap/fonts/ dist/cloud-reader/fonts/ && node ./readium-js/readium-shared-js/readium-cfi-js/node_modules/cpy-cli/cli.js src/css/annotations.css dist/cloud-reader/css/`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the readium-js-viewer@0.24.0 dist:cloudReader:copy script 'node ./readium-js/readium-shared-js/readium-cfi-js/node_modules/mkdirp/bin/cmd.js dist/cloud-reader/scripts/zip && node ./readium-js/readium-shared-js/readium-cfi-js/node_modules/mkdirp/bin/cmd.js dist/cloud-reader/scripts/mathjax && node ./readium-js/readium-shared-js/readium-cfi-js/node_modules/cpy-cli/cli.js src/cloud-reader/index.html dist/cloud-reader/ && node ./readium-js/readium-shared-js/readium-cfi-js/node_modules/cpy-cli/cli.js build-output/z-worker.js dist/cloud-reader/scripts/zip && node ./readium-js/readium-shared-js/readium-cfi-js/node_modules/cpy-cli/cli.js build-output/deflate.js dist/cloud-reader/scripts/zip && node ./readium-js/readium-shared-js/readium-cfi-js/node_modules/cpy-cli/cli.js build-output/inflate.js dist/cloud-reader/scripts/zip && node ./readium-js/readium-shared-js/readium-cfi-js/node_modules/ncp/bin/ncp src/images/ dist/cloud-reader/images/ && node ./readium-js/readium-shared-js/readium-cfi-js/node_modules/cpy-cli/cli.js node_modules/mathjax-single-file/dist/MMLSVG/MathJax.js dist/cloud-reader/scripts/mathjax && node ./readium-js/readium-shared-js/readium-cfi-js/node_modules/ncp/bin/ncp node_modules/bootstrap/fonts/ dist/cloud-reader/fonts/ && node ./readium-js/readium-shared-js/readium-cfi-js/node_modules/cpy-cli/cli.js src/css/annotations.css dist/cloud-reader/css/'.
npm ERR! Make sure you have the latest version of node.js and npm installed.
npm ERR! If you do, this is most likely a problem with the readium-js-viewer package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     node ./readium-js/readium-shared-js/readium-cfi-js/node_modules/mkdirp/bin/cmd.js dist/cloud-reader/scripts/zip && node ./readium-js/readium-shared-js/readium-cfi-js/node_modules/mkdirp/bin/cmd.js dist/cloud-reader/scripts/mathjax && node ./readium-js/readium-shared-js/readium-cfi-js/node_modules/cpy-cli/cli.js src/cloud-reader/index.html dist/cloud-reader/ && node ./readium-js/readium-shared-js/readium-cfi-js/node_modules/cpy-cli/cli.js build-output/z-worker.js dist/cloud-reader/scripts/zip && node ./readium-js/readium-shared-js/readium-cfi-js/node_modules/cpy-cli/cli.js build-output/deflate.js dist/cloud-reader/scripts/zip && node ./readium-js/readium-shared-js/readium-cfi-js/node_modules/cpy-cli/cli.js build-output/inflate.js dist/cloud-reader/scripts/zip && node ./readium-js/readium-shared-js/readium-cfi-js/node_modules/ncp/bin/ncp src/images/ dist/cloud-reader/images/ && node ./readium-js/readium-shared-js/readium-cfi-js/node_modules/cpy-cli/cli.js node_modules/mathjax-single-file/dist/MMLSVG/MathJax.js dist/cloud-reader/scripts/mathjax && node ./readium-js/readium-shared-js/readium-cfi-js/node_modules/ncp/bin/ncp node_modules/bootstrap/fonts/ dist/cloud-reader/fonts/ && node ./readium-js/readium-shared-js/readium-cfi-js/node_modules/cpy-cli/cli.js src/css/annotations.css dist/cloud-reader/css/
npm ERR! You can get information on how to open an issue for this project with:
npm ERR!     npm bugs readium-js-viewer
npm ERR! Or if that isn't available, you can get their info via:
npm ERR!     npm owner ls readium-js-viewer
npm ERR! There is likely additional logging output above.

npm ERR! Please include the following file with any support request:
npm ERR!     C:\Users\Hajar\Projects\ReadiumCloud\test7\readium-js-viewer\npm-debug.log

npm ERR! Windows_NT 10.0.14393
npm ERR! argv "C:\\Program Files\\nodejs\\node.exe" "C:\\Program Files\\nodejs\\node_modules\\npm\\bin\\npm-cli.js" "run" "dist:cloudReader"
npm ERR! node v6.1.0
npm ERR! npm  v3.8.6
npm ERR! code ELIFECYCLE
npm ERR! readium-js-viewer@0.24.0 dist:cloudReader: `npm run dist:cloudReader:copy && npm run dist:cloudReader:css && npm run dist:cloudReader:js`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the readium-js-viewer@0.24.0 dist:cloudReader script 'npm run dist:cloudReader:copy && npm run dist:cloudReader:css && npm run dist:cloudReader:js'.
npm ERR! Make sure you have the latest version of node.js and npm installed.
npm ERR! If you do, this is most likely a problem with the readium-js-viewer package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     npm run dist:cloudReader:copy && npm run dist:cloudReader:css && npm run dist:cloudReader:js
npm ERR! You can get information on how to open an issue for this project with:
npm ERR!     npm bugs readium-js-viewer
npm ERR! Or if that isn't available, you can get their info via:
npm ERR!     npm owner ls readium-js-viewer
npm ERR! There is likely additional logging output above.

npm ERR! Please include the following file with any support request:
npm ERR!     C:\Users\Hajar\Projects\ReadiumCloud\test7\readium-js-viewer\npm-debug.log

 Kind regards,
-Hajar

**Comment by larsvoigt** [comment]
I think ```npm run build``` was run incomplete. The task look for the z-worker.js but is ```ENOENT: no such file or directory,```. Do you have success if you build it without step 4 of your last post?

**Comment by hajarghaem** [comment]
Just to be sure, I run one more time from scratch the following instructions. I have z-worker.js and other required files and folders.
1- git clone --recursive -b develop https://github.com/*readium*/readium-js-viewer.git readium-js-viewer 2- cd readium-js-viewer 
3- and I did other steps in https://github.com/readium/readium-js-viewer/tree/develop#development but not "npm run prepare"
 4- npm run prepare
 6- npm run build
 7- npm run dist 
As you see I did not use your branch (files)! and it is just original Readium and its cloud reader.

**Comment by hajarghaem** [comment]
Using your files:
the "npm run build" does not make "_single-bundle" folder.
the "npm run dist:cloudReader" does not make its required folders and files.

**Comment by hajarghaem** [comment]
Hi @larsvoigt 
I installed npm install epub-full-text-search -g, and can use it, for example "curl -XGET "http://localhost:8085/search?q=math".
I also deployed readium cloud (https://github.com/readium/readium-js-viewer). 
I am going to search for keyword in Readium, by using epub-full-text-search. 
Could you  please let me know what is the best way to integrate as a feature of Readium. 
Thanks in advance

**Comment by hajarghaem** [comment]
Hi @larsvoigt 
I have been working on highlighting text based on CFI.
I could integrate epub-search engin in Readim, as search function.
Now I can extract CFI of a search keyword in json format.
Now I need to highlight my search keyword in the readium. So it is not based on selected text.
The only guidance that I could find is https://docs.google.com/document/d/1O8V0C2XaynobpDApnQvGcfgE4aCDj7BiZ5W-4eOupIk/edit#heading=h.nlvujjtmpl2m. But I could not use it to highlight my searched keyword.
For example my extracted information in jsn format is:
{
    "filename" : "accessible_epub_3",
    "epubTitle" : "Accessible EPUB 3",
    "href" : "ch01.xhtml",
    "baseCfi" : "/6/16[id-id2611884]!",
    "id" : "id-id2611884",
    "cfis" : ["/6/16[id-id2611884]!/4/2[introduction]/4,/1:22,/1:23"]
}
and I tried to work on cloud-reader.
Could you please let me know how should be the right code in order to highlight my example:
I tried to adopt "S.reader.plugins.highlights.addSelectionHighlight(Math.floor(1e6 * Math.random()), "test-highlight")" from
"        r(".icon-annotations").on("click", function () {
            S.reader.plugins.highlights.addSelectionHighlight(Math.floor(1e6 * Math.random()), "test-highlight")
            //S.reader.plugins.highlights.addSelectionHighlight('id-id2611884', "/4,/1:22,/1:23", 123, "highlight");
        });"

But my problem is how do I need to make a link between Math.floor(1e6 * Math.random()) and  'id-id2611884', "/4,/1:22,/1:23", 123,
Or if you could propose me know other ways!

**Comment by larsvoigt** [comment]
Hi @hajarghaem, apologize for my late reply. Did you solve your problem?

**Comment by larsvoigt** [comment]
You can try this:

``` javascript
     /**
         * Highlighter use parts from readium CFI navigation helper class.
         *
         * Own Hightlighter should be full text search module
         * a little bit more independent from future developing waves
         */
        var Highlighter = function (readium) {

            // todo: focus screenreader
            
            var self = this;
            var lastOverlay;

            self.add = function (idref, partialCfi, borderColor) {

                if (lastOverlay)
                    lastOverlay.remove();
                
                var range = readium.reader.getDomRangeFromRangeCfi(new BookmarkData(idref, partialCfi));
                console.log("________________________ range: " + range);

                drawOverlayFromDomRange(range, borderColor)
            };


            function drawOverlayFromDomRange(range, borderColor) {
                var rect = getNodeRangeClientRect(
                    range.startContainer,
                    range.startOffset,
                    range.endContainer,
                    range.endOffset);

                drawOverlayFromRect(rect, borderColor);
            }


            function drawOverlayFromRect(rect, borderColor) {
                var leftOffset, topOffset;

                if (isVerticalWritingMode()) {
                    leftOffset = 0;
                    topOffset = -getPaginationLeftOffset();
                } else {
                    leftOffset = -getPaginationLeftOffset();
                    topOffset = 0;
                }

                addOverlayRect({
                    left: rect.left + leftOffset,
                    top: rect.top + topOffset,
                    width: rect.width,
                    height: rect.height
                }, borderColor, self.getRootDocument());
            }


            function addOverlayRect(rects, borderColor, doc) {

                if (!(rects instanceof Array)) {
                    rects = [rects];
                }
                for (var i = 0; i != rects.length; i++) {
                    var rect = rects[i];
                    var overlayDiv = doc.createElement('div');
                    overlayDiv.style.position = 'absolute';
                    $(overlayDiv).css('z-index', '1000');
                    $(overlayDiv).css('pointer-events', 'none');
                    $(overlayDiv).css('opacity', '0.4');
                    overlayDiv.style.border = '2px dashed ' + borderColor;
                    overlayDiv.style.background = 'yellow';
                    overlayDiv.style.margin = overlayDiv.style.padding = '0';
                    overlayDiv.style.top = (rect.top ) + 'px';
                    overlayDiv.style.left = (rect.left ) + 'px';
                    // we want rect.width to be the border width, so content width is 2px less.
                    overlayDiv.style.width = (rect.width - 2) + 'px';
                    overlayDiv.style.height = (rect.height - 2) + 'px';
                    doc.documentElement.appendChild(overlayDiv);
                    lastOverlay = overlayDiv;
                }
            }

            function getPaginationLeftOffset() {

                var $htmlElement = $("html", self.getRootDocument());
                var offsetLeftPixels = $htmlElement.css(isVerticalWritingMode() ? "top" : (isPageProgressionRightToLeft() ? "right" : "left"));
                var offsetLeft = parseInt(offsetLeftPixels.replace("px", ""));
                if (isNaN(offsetLeft)) {
                    //for fixed layouts, $htmlElement.css("left") has no numerical value
                    offsetLeft = 0;
                }
                if (isPageProgressionRightToLeft() && !isVerticalWritingMode()) return -offsetLeft;
                return offsetLeft;
            }

            function getNodeRangeClientRect(startNode, startOffset, endNode, endOffset) {
                var range = createRange();
                range.setStart(startNode, startOffset ? startOffset : 0);
                if (endNode.nodeType === Node.ELEMENT_NODE) {
                    range.setEnd(endNode, endOffset ? endOffset : endNode.childNodes.length);
                } else if (endNode.nodeType === Node.TEXT_NODE) {
                    range.setEnd(endNode, endOffset ? endOffset : 0);
                }
                return normalizeRectangle(range.getBoundingClientRect(), 0, 0);
            }

            function normalizeRectangle(textRect, leftOffset, topOffset) {

                var plainRectObject = {
                    left: textRect.left,
                    right: textRect.right,
                    top: textRect.top,
                    bottom: textRect.bottom,
                    width: textRect.right - textRect.left,
                    height: textRect.bottom - textRect.top
                };
                offsetRectangle(plainRectObject, leftOffset, topOffset);
                return plainRectObject;
            }

            function offsetRectangle(rect, leftOffset, topOffset) {

                rect.left += leftOffset;
                rect.right += leftOffset;
                rect.top += topOffset;
                rect.bottom += topOffset;
            }

            function isPageProgressionRightToLeft() {
                return readium.reader.getPaginationInfo().rightToLeft;
            }

            function isVerticalWritingMode() {
                return readium.reader.getPaginationInfo().isVerticalWritingMode;
            }

            this.getRootDocument = function () {
                return $('#epubContentIframe')[0].contentDocument;
            };

            function createRange() {
                return self.getRootDocument().createRange();
            }
        };
```
and call it after the spine item is loaded!

``` javascript 
const  highlighter = new Highlighter(readium);
try {
  highlighter.add(idref, partialCfi, 'blue');
 } catch (e) {
   console.error(e);
}
```

**Comment by larsvoigt** [comment]
If you call this with your cfi.  
```javascript
readium.reader.openSpineItemElementCfi(...); 
```
Does it open the right page?

**Comment by hajarghaem** [comment]
Hi @larsvoigt ,
Thanks! I just noticed your comments. I will work on that and update you.

**Comment by hajarghaem** [comment]
Hi @larsvoigt,
I used other way to highlight my search keyword using epub-search in readium.
There is an small issue about my extracted CFI from epub-search. IN CFI.js: The  startOffset and endOffset does not contains the length of the queried key word. For example I search for word=test. then it is an example CFI that epub-search makes: CFI=" /4/2/10/48,/1:237,/1:238"
You can see that the startOffset=237 and the endOffset= 238.
The word.length=4 so I expected to have CFI=" /4/2/10/48,/1:235,/1:239" , an startOffset=235 and an endOffset= 239 (according to the book I search on it!). 
Could you please provide me more information.
Thanks in advance

**Comment by hajarghaem** [comment]
Hi @larsvoigt, Is that possible to provide me some information about my previous comment?
Thanks in advance

**Comment by larsvoigt** [comment]
Hi @hajarghaem, 
at the moment it is difficult to reproduce this for me. Can you share your document with me lars.voigt@dzb.de? Or better, it would be nice to have any test case for this.  Feel free to define [here]( https://github.com/larsvoigt/epub-full-text-search/blob/master/tests/specs/test_search.js)  a new test case. 
Best Lars

**Comment by hajarghaem** [comment]
Hi @larsvoigt 
I fixed it by adding
 endOffset= startOffset + query[0][0].length; instead of endOffset= startOffset + query[0].length;
 in line 81 https://github.com/larsvoigt/epub-full-text-search/blob/master/src/CFI.js

**Comment by hajarghaem** [comment]
Hi @larsvoigt 
I was wondering why some results of my local search a word (e.g. "data") in the same book "accessible_epub_3" locally using epub-search is different from what I retrieve from http://protected-dusk-3051.herokuapp.com/?searchbox=data
For example searching for "data" in "Chapter 1. Introduction" locally, shows the following results:
{
		"filename" : "accessible_epub_3",
		"epubTitle" : "Accessible EPUB 3",
		"href" : "ch01.xhtml",
		"baseCfi" : "/6/16[id-id2611884]!",
		"id" : "id-id2611884",
		"cfis" : ["/4/2[introduction]/4,/1:377,/1:381", "/4/2[introduction]/6,/1:309,/1:313", "/4/2[introduction]/8,/1:59,/1:63", "/4/2[introduction]/12,/1:11,/1:15"]
	},
But  http://protected-dusk-3051.herokuapp.com/?searchbox=data does not provide these information "id-id2611884"?

**Comment by hajarghaem** [comment]
@larsvoigt 
I see that epub-search finds just some of existing keywords in some cases and not all existing keywords. 
For example in book="accessible_epub_3", file="ch01s02.xhtml", **there are 5** keyword="someone".
But the search result finds **just 3 of them**: (I use: http://localhost:8085/search?q=someone&t=Accessible+EPUB+3)
{
		"filename" : "accessible_epub_3",
		"epubTitle" : "Accessible EPUB 3",
		"href" : "ch01s02.xhtml",
		"baseCfi" : "/6/18[id-id2627310]!",
		"id" : "id-id2627310",
		"cfis" : ["/4/2/22/2/2,/1:1,/1:8", "/4/2/22/4/2,/1:1,/1:8", "/4/2/22/6/2,/1:1,/1:8"]
	}

I am going to fix this bug. 
Could you please let me know your thought?
Thanks in advance

**Comment by larsvoigt** [comment]
Hi @hajarghaem, great to all! Is it possible you can make pull requests for your fixing?  That would be very nice. And [test cases]( https://github.com/larsvoigt/epub-full-text-search/blob/master/tests/specs/test_search.js) are very importent because we can see if your fixings have influence of the rest of the code base  and the isssues are better reproducible.  
Best Lars

**Comment by hajarghaem** [comment]
Hi @larsvoigt Thanks! I will make the test case and pull request for my fixing.
Is that possible for you to answer my previous comment? 
Thanks in advance

**Comment by larsvoigt** [comment]
Hi @hajarghaem, you are right. It should return 5 hits for the keyword "someone". That is not correct ...

**Comment by hajarghaem** [comment]
Hi @larsvoigt 
Thanks for your time ad response, could you please let me know where (which files?) I can start to fix it.

**Comment by larsvoigt** [comment]
[getElementsThatContainsQuery()](https://github.com/larsvoigt/epub-full-text-search/blob/master/src/CFI.js#L73). Oh I think the problem is if the query will match several times in the same text element but the current algorithm will only record one matching per element. It needs a loop to get all matchings within the text element  ``` $(this).text() ```.

**Comment by hajarghaem** [comment]
@larsvoigt 
thanks I will consider it and will update you!

**Comment by larsvoigt** [comment]
I think this issue is outdated so I close it.

**Comment by rudra0713** [comment]
@hajarghaem ,
Can you please tell me how do I incorporate epub-full-text-search in my readium project ? 
According to the section library use ,first command is 'import epubSearch from epub-full-text-search'. Import is a ES6 syntax. So I transpiled it using babel and this line converts to this -

"use strict";

var _epubFullTextSearch = require("epub-full-text-search");

var _epubFullTextSearch2 = _interopRequireDefault(_epubFullTextSearch);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

When I build and run and add the dist folder in chrome extension, in console I get the error 
"no epub-full-text-search" .  Seems to me that require has already been defined in Readium and this is not compatible with javascript require().

**Comment by hajarghaem** [comment]
Hi @rudra0713,

You should first run your epub-full-text-search(node.js) as a separate server (you can test it using http://localhost:8085/search?q=epub&t=Accessible+EPUB+3, please do all steps that are mentioned in https://github.com/larsvoigt/epub-full-text-search. For eaxmple you must first index books ), and then run your readium-cloud-reader(in node.js please follow all steps that are mentioned in https://github.com/readium/readium-js-viewer regarding to "Cloud reader deployment"). Then, you must define an access link to "epub-full-text-search(node.js)" withon your readium-cloud-reader (inside readium-js-viewer/src/js/EpubReader7.js). If yo ant to access to epub-full-text-search within readium-cloud-reader, then it is better to make an icon-search for it somewhere in the navigation bar of readium (inside readium-js-viewer/src/templates/reader-navbar.html ). Then using an ajax call you can access to epub-full-text-search. I ut some part of my code here but it was for a couple of months ago, 
$(".icon-search").on("click", function () {
			var type="test-highlight";
			var element = document.getElementById("app-navbar");
			if(document.getElementById("search-id")){
				var child=document.getElementById("search-id");
				element.removeChild(child);
				readium.reader.plugins.highlights.removeHighlightsByType(type);
			}; 
			var type="test-highlight";
			readium.reader.plugins.highlights.removeHighlightsByType(type);
			if (document.getElementById('readium-toc-body').innerHTML != ""){document.getElementById('readium-toc-body').innerHTML = "";};
			window.myValue="test2";
			
			var url = "http://localhost:8085/search?q=epub"; 
			var x = document.createElement("input");
			x.setAttribute("style","float:left");
			x.setAttribute("id","search-id");
			var foo = document.getElementById("app-navbar");
			foo.appendChild(x);
			
			
			
			var getJSON = function(url) {
				return new Promise(function(resolve, reject) {
					var xhr = new XMLHttpRequest();
					xhr.open('get', url, true);
					xhr.responseType = 'json';
					xhr.onload = function() {
						var status = xhr.status;
						if (status == 200) {
							resolve(xhr.response);
						} else {
							reject(status);
						}
					};//alert(url);
					xhr.send();
				});
			}; 
			document.getElementById("search-id").addEventListener("change", function() {
				
				var currentBookTitle = /[^/]*$/.exec(ebookURL)[0];//alert(currentBookTitle);
				var currentBookTitle = currentBookTitle.replace(/%20/g, "");//alert(currentBookTitle);
				currentBookTitle = currentBookTitle.replace(/[_-]/g, "");//alert(currentBookTitle);
				getJSON('http://localhost:8085/search?q=' + x.value+ '&t=' + currentBookTitle ).then(function(data) {
					window.searchKeyWord=x.value;//alert('x.value='+x.value);
					window.allData=data;
					if (window.allData==''){window.myValue="test3";};
					var indexedBookTitle="";
					var i=0 , j=0 , filenum=0;
					 var filehref=['name'];
					  for (i = 0; i < data.length ; i++) {
						indexedFileHref = JSON.stringify(data[i].href);//alert(data[i].href);


My codes are just for inspiration :).

Kind regards,
Hajar

**Comment by rudra0713** [comment]
Hi @hajarghaem , thank you very much for your extraordinary help.  I am still in testing phase but I am achieving good result until now. However, right now, first I have to run the server for searching at say port 8085 and then make ajax call to read data from my readium project.  This is not quite suitable for development purpose i think. I was thinking of writing a script that when executed, will run the server for search in epub. My goal is to ensure that when a user tries to use this project, he / she only has to first run the script and then run the readium project. Do you think this is possible ?
regards 
Rudra

**Comment by hajarghaem** [comment]
Hi @rudra0713,
You are welcome.
What I do is:
1- run epub-search in 8085, and then readium in 8080
2- I start reading a book via readium and if I want to search for a word such as "epub", then I click on search button that I added to readium navigation bar.
3- then based on your result visualization methods you can see your result somewhere (e.g., you can set it to view your result in Table-Of-Contents, or in a dialogbox window,.....)
4- you can use CFI of your results (that can be access via epub-search server output) in order to go to the specefic page contains your specefic search result.
You can use "readium.reader.openSpineItemElementCfi(spanId,spanCfi)"

I prefer to run "epub-search" as a separate server and not as part of readium, becasue if there is an update for "eapub-search" then I may need some modification in way of access to the result of it in my readium.

I know it is not the answer of your question but I hope i could turn on some lights.
Regards,
Hajar

**Comment by rudra0713** [comment]
In order to run epub search ,What I do is 
a. open my terminal / cmd
b. write two commands -
           epub-search start
           epub-search writeToIndex -p {path}
  as mentioned by Lars.
I think that first running two commands from terminal and then my readium project is a bit lengthy.   I use ajax call to fetch data from epub-search inside readium . I was thinking of compacting these procedures. I understand your reasoning But if I were to complete this project and give it to someone, then he/she would have to first install full-text-search and then run those commands which was not even quite easy for me.

**Comment by hajarghaem** [comment]
that is right for oflinne usage. But I use readium-cloud online so my services are online and the user just need to use my servies online and no need for any installation for the user.

**Comment by larsvoigt** [comment]
Hi @rudra0713, @hajarghaem. Did you resolve your integration issues?

**Comment by hajarghaem** [comment]
@larsvoigt I have no problem as I have worked online. I have no idea for offline usage.

**Comment by larsvoigt** [comment]
For testing with readium cloud reader you can do it like this: 
```javascript
function search() {
    ...
    const request = host + `/search?q=${term}&t=${epubTitle}`;
    $.getJSON(request, '', {})
        .done(hits => {
            if (hits.length > 0) {
                for (var i in hits) {
                    for (var ii in hits[i].cfis) {
                        const idref = hits[i].id;
                        const elementcfi = hits[i].cfis[ii].cfi.split('!')[1];
                        const goto = `&goto={"idref" : "${idref}", "elementCfi" : "${elementcfi}"}`;
                        const epub = remoteEPUB;
                        const gotoLink = host + `/?epub=${epub}${goto}`;

                        // add goto link to user hit list 
                        // open goto link onClick() hit list ...
                    }
                }
            }
        }
   }
```
Example [Goto-Link](https://readium.firebaseapp.com/?epub=epub_content%2Faccessible_epub_3&goto=%7B%22idref%22%3A%22id-id2442754%22%2C%22elementCfi%22%3A%22%2F4%2F2%5BI_book_d1e1%5D%2F12%2F2%5Bid2603834%5D%2F2%2C%2F3%3A20%2C%2F3%3A31%22%7D)

or write an own module for the readium viewer.

**Comment by larsvoigt** [comment]
@hajarghaem Offline is a little bit more difficult :-).  Migrate your cloud reader to an Electron App? Webworkers? I have no really an idea ...

**Comment by larsvoigt** [comment]
See also the new [DEMO](http://protected-dusk-3051.herokuapp.com/) with the new on-the-fly indexing feature ...

**Comment by hajarghaem** [comment]
@larsvoigt that is impressive good job :)
